sys/congure/quic/congure_quic.c

```diff
@@ -0,0 +1,292 @@
+ * (for pacing calculation) as basis for this implementation.
+ */
+
+#include <assert.h>
+#include <limits.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "clist.h"
+#include "timex.h"
+#include "ztimer.h"
+
+#include "congure/quic.h"
+
+static void _snd_init(congure_snd_t *cong, void *ctx);
+static int32_t _snd_inter_msg_interval(congure_snd_t *cong, unsigned msg_size);
+static void _snd_report_msg_sent(congure_snd_t *cong, unsigned sent_size);
+static void _snd_report_msg_discarded(congure_snd_t *cong, unsigned msg_size);
+static void _snd_report_msgs_lost(congure_snd_t *cong, congure_snd_msg_t *msgs);
+static void _snd_report_msg_acked(congure_snd_t *cong, congure_snd_msg_t *msg,
+                                  congure_snd_ack_t *ack);
+static void _snd_report_ecn_ce(congure_snd_t *cong, ztimer_now_t time);
+
+static const congure_snd_driver_t _driver = {
+    .init = _snd_init,
+    .inter_msg_interval = _snd_inter_msg_interval,
+    .report_msg_sent = _snd_report_msg_sent,
+    .report_msg_discarded = _snd_report_msg_discarded,
+    .report_msgs_timeout = _snd_report_msgs_lost,
+    .report_msgs_lost = _snd_report_msgs_lost,
+    .report_msg_acked = _snd_report_msg_acked,
+    .report_ecn_ce = _snd_report_ecn_ce,
+};
+
+static inline bool _in_recov(congure_quic_snd_t *c, ztimer_now_t sent_time)
+{
+    return sent_time <= c->recovery_start;
+}
+
+static void _on_congestion_event(congure_quic_snd_t *c, ztimer_now_t sent_time)
+{
+    if (_in_recov(c, sent_time)) {
+        return;
+    }
+    /* enter congestion recovery period */
+    c->recovery_start = ztimer_now(ZTIMER_MSEC);
+    c->ssthresh = (c->super.cwnd * c->consts->loss_reduction_numerator)
+                / c->consts->loss_reduction_denominator;
+    c->super.cwnd = (c->ssthresh > c->consts->min_wnd)
+                    ? c->ssthresh : c->consts->min_wnd;
+    if (c->consts->cong_event_cb) {
+        c->consts->cong_event_cb(c->super.ctx);
+    }
+}
+
+static void _update_rtts(congure_quic_snd_t *c, ztimer_now_t msg_send_time,
+                         ztimer_now_t ack_recv_time, uint16_t ack_delay)
+{
+    uint16_t latest_rtt;
+
+    assert((ack_recv_time - msg_send_time) <= UINT16_MAX);
+    /* we assume that is in the uint16_t range, but just in case NDEBUG
+     * is set, let's cap it at UINT16_MAX */
+    if ((ack_recv_time - msg_send_time) > UINT16_MAX) {
+        latest_rtt = UINT16_MAX;
+    }
+    else {
+        latest_rtt = ack_recv_time - msg_send_time;
+    }
+
+    if (c->first_rtt_sample > 0) {  /* an RTT sample was taken */
+        c->min_rtt = (c->min_rtt > latest_rtt) ? latest_rtt : c->min_rtt;
+        /* adjust latest_rtt for ack_delay if plausible */
+        if (latest_rtt > (c->min_rtt + ack_delay)) {
+            latest_rtt -= ack_delay;
+        }
+        c->rtt_var = ((3U * c->rtt_var) / 4U)
+                   + (abs(((int)c->smoothed_rtt) - latest_rtt) / 4U);
```

Comment:
```suggestion
                   + (abs((int)c->smoothed_rtt - (int)latest_rtt) / 4U);
```
not sure why this is only an issue on MSP430